python/tk_multi_launchapp/launch_apps.py

```diff
@@ -0,0 +1,651 @@
+                # the current environment.
+                self._tk_app.log_warning(
+                    "%s is not loaded in the current environment." % app_engine
+                )
+                continue
+
+            # If there are Project restrictions, check if the current Project is one of them.
+            app_project_ids = [proj["id"] for proj in app_projects]
+            if app_project_ids and (not ctx_project or ctx_project["id"] not in app_project_ids):
+                self._tk_app.log_debug(
+                    "Context Project %s not found in Project restrictions for app [%s]." %
+                    (ctx_project, app_menu_name)
+                )
+                continue
+
+            # If there are user restrictions, check if the current user is one of them.
+            app_user_ids = [user["id"] for user in app_users]
+            if app_user_ids and (not ctx_user or ctx_user["id"] not in app_user_ids):
+                self._tk_app.log_debug(
+                    "Context user %s not found in user restrictions for app [%s]." %
+                    (ctx_user, app_menu_name)
+                )
+                continue
+
+            # If there are group restrictions, check if the current user is in one of them.
+            app_group_ids = [grp["id"] for grp in app_groups]
+            usr_group_ids = _sg_user_group_ids(self._tk_app.shotgun, ctx_user)
+            if app_group_ids and (not user_group_ids or
+                    not set(user_group_ids).intersection(app_group_ids)):
+                self._tk_app.log_debug(
+                    "Context user %s is not a member of Group restrictions for app [%s]." %
+                    (ctx_user, app_menu_name)
+                )
+                continue
+
+            # Download the thumbnail to use as the app's icon.
+            if app_icon:
+                sg_icon = shotgun_data.ShotgunDataRetriever.download_thumbnail(
+                    app_icon, self._tk_app
+                )
+                app_icon = sg_icon
+                self._tk_app.log_debug("App icon from ShotgunDataRetriever : %s" % app_icon)
+
+            # Parse the Software versions field to determine the specific list of
+            # versions to load. Assume the list of versions is stored as comma-separated
+            # in SG.
+            app_versions = sw_app["sg_versions"]
+            if app_versions:
+                for app_version in [v.strip() for v in app_versions.split(",")]:
+                    if not app_version :
+                        continue
+                    self._init_launch_command(
+                        app_menu_name, app_icon, app_engine, app_path, app_args, app_version
+                    )
+            else:
+                self._init_launch_command(
+                    app_menu_name, app_icon, app_engine, app_path, app_args
+                )
+
+
+def _get_clean_version_string(version):
```

Comment:
as suggested above, I think all of these helper methods can be moved into a `util.py`